### PR TITLE
feat(ui): add search filtering to recent folders overlay (Cmd+O)

### LINKED ISSUE #274: [Feature]: Searchable recent folders overlay (Cmd+O)
## Status Quo

The Cmd+O recent folders overlay (`src/ui/components/recent_folders_overlay.zig`) displays up to 10 most-visited directories. The user can navigate with arrow keys or select with Cmd+1–9, but there is no way to filter the list by typing. For users with many visited directories, finding a specific one requires scanning the list visually.

Two other overlays — Story (`story_overlay.zig`) and Reader (`reader_overlay.zig`) — already implement nearly identical search: case-insensitive substring matching, `search_query: ArrayList(u8)`, `SearchMatch` struct, same highlight colors/alpha, same keyboard handling, and same search bar rendering. This logic is currently duplicated between them with no shared utility. Additionally, Reader overlay's search highlight positioning is misaligned — highlights render to the right of the actual matched text, while Story overlay's highlights are positioned correctly.

## Objectives

Let users quickly narrow down the recent folders list by typing a substring. The search should feel instant and visually consistent with existing search in Story and Reader overlays. As part of this work, extract the duplicated search logic from Story and Reader overlays into a shared module, then reuse it in all three overlays — using Story overlay's correct highlight positioning as the reference implementation and fixing the Reader overlay's misalignment in the process.

## User Flow

**Trigger:** User opens the overlay with Cmd+O.

1. User starts typing — the overlay consumes keyboard input and filters the full folder history to show only directories whose path contains the typed substring (case-insensitive), capped at 10 results.
2. The matched substring is highlighted within each entry using the same accent/selection colors as Story and Reader search.
3. User presses Backspace to remove the last character of the query; the list re-filters.
4. User navigates filtered results with arrow keys or Cmd+1–9, then presses Return to select.
5. Closing the overlay (Escape or selecting a folder) clears the search query.

**Result:** The user navigates to the desired directory without scanning through an unfiltered list.

## Scope

**In scope:**

- Extract shared search utilities (case-insensitive find, keyboard input handling, search bar rendering) from the duplicated Story/Reader code into a reusable module
- Refactor Story and Reader overlays to use the shared module (no visual or behavioral change for Story; fixes highlight misalignment for Reader)
- Text input consumption when the recent folders overlay is open (append characters, Backspace to delete)
- Case-insensitive substring filtering across the full folder history, showing up to 10 matches
- Substring highlight rendering on matching entries (reuse accent/selection color scheme)
- Search query display (visual indicator of what the user has typed, consistent with existing search bars)
- Clearing query on overlay close

**Out of scope:**

- Changing how folders are loaded or ranked (that's #273)
- Cmd+F activation — search starts immediately on typing, no separate toggle needed
- Next/previous match navigation (the list is short enough that arrow keys suffice)
- Regex or fuzzy matching — substring only

### Affected Modules

- `src/ui/components/` (new file): Shared search utility module (e.g. `search_utils.zig`) — case-insensitive find, search bar rendering helpers, keyboard input handling
- `src/ui/components/story_overlay.zig`: Refactor to use shared search module
- `src/ui/components/reader_overlay.zig`: Refactor to use shared search module; fix highlight positioning bug
- `src/ui/components/recent_folders_overlay.zig`: Add search query state, keyboard input handling, filtering logic, and match highlight rendering

### Tasks

Each task is independently testable. Ordered by implementation sequence.

1. [ ] Extract shared search utilities from Story and Reader overlays into a new module (e.g. `src/ui/components/search_utils.zig`) — `findCaseInsensitive`, search bar rendering, keyboard input append/backspace logic. Use Story overlay's highlight positioning (`byteOffsetToPixelX`) as the reference implementation, since Reader overlay's approach (fixed `char_w` multiplication) produces misaligned highlights.
2. [ ] Refactor `story_overlay.zig` to use the shared search module instead of its inline implementation
3. [ ] Refactor `reader_overlay.zig` to use the shared search module instead of its inline implementation — this should fix the highlight misalignment bug by adopting the correct positioning logic
4. [ ] Add search state to `RecentFoldersOverlayComponent` — `search_query: ArrayList(u8)`, filtered results list
5. [ ] Implement keyboard input handling — consume `SDL_EVENT_TEXT_INPUT` when overlay is open, Backspace support, clear on close
6. [ ] Implement filtering — on each query change, re-filter full folder history by case-insensitive substring, cap at 10 results, update display
7. [ ] Implement substring highlight rendering on filtered entries using shared highlight style
8. [ ] Render search query indicator (search bar) consistent with Story/Reader overlays
9. [ ] Write tests: substring matching, empty query shows unfiltered list, backspace behavior, case insensitivity, cap at 10
10. [ ] Update `README.md` — mention search in the Recent Folders section
11. [ ] Update `docs/ARCHITECTURE.md` — document the new shared search utility module

### New Dependencies

None

## Acceptance Criteria

- [ ] All tasks completed
- [ ] Typing in the open overlay filters the folder list by substring
- [ ] Matched substrings are visually highlighted with the same style as Story/Reader search
- [ ] Backspace removes the last character; empty query shows the default unfiltered list
- [ ] Cmd+1–9 and arrow keys still work on filtered results
- [ ] Story and Reader overlays use the extracted shared module with no visual or behavioral change (except Reader's highlight positioning is now correct)
- [ ] Tests cover happy path and edge cases
- [ ] `zig build`, `zig build test`, and `just lint` pass
- [ ] Documentation updated

### PR DESCRIPTION
**Pull request by forketyfork**
## Summary

Closes #274

The Cmd+O recent folders overlay now supports instant search filtering — start typing to narrow the list by substring, with matched text highlighted. Search activates on keypress (no Cmd+F needed), Backspace edits the query, and closing the overlay resets it.

As part of this work, duplicated search logic from the Story and Reader overlays was extracted into a shared `search_utils.zig` module. Both overlays now use it with no behavioral change, except that Reader's search highlight positioning is fixed — it previously used a fixed character-width multiplication that produced misaligned highlights on variable-width or multi-byte text.

### What changed

**New shared module (`search_utils.zig`)** provides `findCaseInsensitive`, `rebuildMatches`, `renderSearchBar`, `makeTextTexture`, and shared types (`SearchMatch`, `TextTex`). Includes 15 unit tests covering edge cases (empty queries, case insensitivity, multiple matches per line, offset handling, match clamping).

**Story overlay** now delegates to search_utils for match rebuilding, search bar rendering, and text texture creation. No visual or behavioral change.

**Reader overlay** same delegation, plus the highlight rendering was rewritten to walk through text runs with actual font measurement (`byteOffsetToPixelX` + `measureScaledTextWidth`) instead of the old `char_w * byte_offset` approach. This fixes the highlight misalignment bug.

**Recent folders overlay** gained search state (`search_query`, `filtered_indices`), text input handling, per-keystroke refiltering across the full folder history (up to 1000 entries from persistence, displayed results capped at 10), a search bar rendered below the title, and substring highlight rendering on path entries. Truncated paths (with `...` prefix) are handled correctly — highlights match against the displayed text, not the original path.

## Test plan

- [ ] Open Cmd+O, type a substring — list filters to matching directories only
- [ ] Verify matched substrings are highlighted with accent color on each entry
- [ ] Press Backspace to shorten the query — list re-expands; empty query shows the default unfiltered list
- [ ] Arrow keys and Cmd+1-9 work correctly on the filtered results
- [ ] Press Escape or select a folder — overlay closes and query resets
- [ ] Open Story overlay, use Cmd+F search — verify highlights still align correctly (no regression)
- [ ] Open Reader mode (Cmd+R), use Cmd+F search — verify highlights now align with the actual rendered text (previously misaligned)
- [ ] Test with a long path that gets truncated with `...` — highlight should still be positioned correctly on the visible portion

### CODE REVIEW

File: src/ui/components/story_overlay.zig [hunk 1]
```diff
@@ -191,55 +184,12 @@ pub const StoryOverlayComponent = struct {
...
-        if (self.selected_match) |idx| {
-            if (idx >= self.matches.items.len) {
-                self.selected_match = 0;
-            }
-        } else {
-            self.selected_match = 0;
+        const plain_texts = self.allocator.alloc([]const u8, self.lines.items.len) catch return;
+        defer self.allocator.free(plain_texts);
+        for (self.lines.items, 0..) |line, i| {
+            plain_texts[i] = line.plain_text;
```
[the remark this answers is not in the document]
Comment (forketyfork) [reply src/ui/components/story_overlay.zig]: Fixed — blanking out `.blank` and `.horizontal_rule` entries when building the `plain_texts` slice, restoring the pre-refactor skip behavior.

File: src/ui/components/reader_overlay.zig [hunk 2]
```diff
@@ -280,55 +273,12 @@ pub const ReaderOverlayComponent = struct {
...
-        if (self.selected_match) |idx| {
-            if (idx >= self.matches.items.len) {
-                self.selected_match = 0;
-            }
-        } else {
-            self.selected_match = 0;
+        const plain_texts = self.allocator.alloc([]const u8, self.lines.items.len) catch return;
+        defer self.allocator.free(plain_texts);
+        for (self.lines.items, 0..) |line, i| {
+            plain_texts[i] = line.plain_text;
```
[the remark this answers is not in the document]
Comment (forketyfork) [reply src/ui/components/reader_overlay.zig]: Fixed — blanking out `.blank`, `.horizontal_rule`, and `.prompt_separator` entries when building the `plain_texts` slice, restoring the pre-refactor skip behavior.

File: src/ui/components/recent_folders_overlay.zig [hunk 3]
```diff
@@ -207,39 +248,55 @@ pub const RecentFoldersOverlayComponent = struct {
...
-                            if (digit_idx < self.folders.items.len) {
-                                self.emitChangeDir(actions, host.focused_session, self.folders.items[digit_idx].abs_path);
-                                self.overlay.startCollapsing(host.now_ms);
+                            if (self.filteredFolder(digit_idx)) |folder| {
+                                self.emitChangeDir(actions, host.focused_session, folder.abs_path);
+                                self.closeOverlay(host.now_ms);
                                 return true;
                             }
                         }
                     }
```
[the remark this answers is not in the document]
Comment (forketyfork) [reply src/ui/components/recent_folders_overlay.zig:275]: Fixed — added `return true` at the end of the keydown handler block to consume all unhandled keydowns while the overlay is open.

File: src/ui/components/recent_folders_overlay.zig [hunk 4]
```diff
@@ -317,7 +374,7 @@ pub const RecentFoldersOverlayComponent = struct {
         const font_size = dpi.scale(@max(12, @min(20, @divFloor(rect.h, 2))), ui_scale);
         const fonts = cache.get(font_size) catch return;
 
-        const glyph = "⌘O";
+        const glyph = "\xe2\x8c\x98O";
```
[the remark this answers is not in the document]
Comment (forketyfork) [reply src/ui/components/recent_folders_overlay.zig]: Reverted both occurrences to the literal `⌘` — `zig fmt` preserves it fine, the byte escapes were unnecessary.
